File: android/MicroService/src/main/java/org/dweb_browser/microservice/ipc/NativePort.kt
package org.dweb_browser.microservice.ipc

import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.channels.Channel
import org.dweb_browser.helper.Callback
import org.dweb_browser.helper.Signal
import org.dweb_browser.helper.SimpleCallback
import org.dweb_browser.helper.SimpleSignal
import java.util.concurrent.atomic.AtomicInteger

class NativeMessageChannel<T1, T2> {
  /**
   * 默认锁住，当它解锁的时候，意味着通道关闭
   */
  private val closePo = SharedCloseSignal()
  private val channel1 = Channel<T1>()
  private val channel2 = Channel<T2>()
  val port1 = NativePort(channel1, channel2, closePo)
  val port2 = NativePort(channel2, channel1, closePo)
}

class SharedCloseSignal {
  private val closeSignal = SimpleSignal()
  fun onClose(cb: Callback<Unit>) = closeSignal.listen(cb)
  private var closed = false
   fun isClosed(): Boolean {
    synchronized(closeSignal) {
      return closed
    }
  }

   suspend fun emitClose() {
     synchronized(closeSignal) {
       if (closed) {
         return
       }
       closed = true
     }
     closeSignal.emitAndClear()
  }
}

class NativePort<I, O>(
  private val channel_in: Channel<I>,
  private val channel_out: Channel<O>,
  private val cs: SharedCloseSignal
) {
  companion object {
    private var uid_acc = AtomicInteger(1);
  }

  private val uid = uid_acc.getAndAdd(1)
  override fun toString() = "#p$uid"

  private var started = false
  suspend fun start() {
    if (started) return else started = true
    /**
     * 等待 close 信号被发出，那么就关闭出口、触发事件
     */
    if (cs.isClosed()) {
      _close()
    } else {
      cs.onClose {
        _close()
      }
    }

    debugNativeIpc("port-message-start/$this")
    for (message in channel_in) {
      debugNativeIpc("port-message-in/$this << $message")
      _messageSignal.emit(message)
      debugNativeIpc("port-message-waiting/$this")
    }
    debugNativeIpc("port-message-end/$this")
  }

  private val _closeSignal = SimpleSignal()

  fun onClose(cb: SimpleCallback) = _closeSignal.listen(cb)

  suspend fun close() {
    return cs.emitClose()
  }

  private suspend fun _close() {
    /// 关闭输出就行了
    channel_out.close()
    _closeSignal.emitAndClear()
    debugNativeIpc("port-closed/${this}")
  }

  private val _messageSignal = Signal<I>()

  /**
   * 发送消息，这个默认会阻塞
   */
  @OptIn(DelicateCoroutinesApi::class)
  suspend fun postMessage(msg: O) {
    debugNativeIpc("message-out/$this >>","$msg ${!channel_out.isClosedForSend}")
    if (!channel_out.isClosedForSend) {
      channel_out.send(msg)
    } else {
      debugNativeIpc("postMessage", " handle the closed channel case!")
    }
  }

  /**
   * 监听消息
   */
  fun onMessage(cb: Callback<I>) = _messageSignal.listen(cb)
}

File: next/dweb-browser/MicroService/src/Ipc/NativePort.cs
using System.Threading.Tasks.Dataflow;

namespace DwebBrowser.MicroService;

public class SharedCloseSignal
{
    private readonly HashSet<Signal> CloseSignal = new();
    public event Signal OnClose
    {
        add { if (value != null) lock (CloseSignal) { CloseSignal.Add(value); } }
        remove { lock (CloseSignal) { CloseSignal.Remove(value); } }
    }
    private bool Closed = false;
    public bool isClosed()
    {
        lock (CloseSignal)
        {
            return Closed;
        }
    }
    public Task EmitClose()
    {
        lock (CloseSignal)
        {
            if (Closed)
            {
                return Task.CompletedTask;
            }
            Closed = true;
            return CloseSignal.EmitAndClear();
        }
    }
}

public class NativePort<I, O>
{
    static readonly Debugger Console = new("NativePort");
    private BufferBlock<I> _channel_in { get; set; }
    private BufferBlock<O> _channel_out { get; set; }
    private SharedCloseSignal _cs { init; get; }

    public NativePort(BufferBlock<I> channel_in, BufferBlock<O> channel_out, SharedCloseSignal cs)
    {
        _channel_in = channel_in;
        _channel_out = channel_out;
        _cs = cs;
    }

    private static int s_uid_acc = 0;

    private readonly int _uid = Interlocked.Increment(ref s_uid_acc);

    public override string ToString() => string.Format("#p{0}", _uid);

    private bool _started { get; set; } = false;

    public async Task StartAsync()
    {
        if (_started)
        {
            return;
        }
        else
        {
            _started = true;
        }

        if (_cs.isClosed())
        {
            await _close();
        }
        else
        {
            _cs.OnClose += (_) => _close();
        }

        Console.Log("Start", "port-message-start/{0}", this);
        await foreach (I message in _channel_in.ReceiveAllAsync())
        {
            Console.Log("Start", "port-message-in/{0} {1}", this, message);
            await (MessageSignal.Emit(message)).ForAwait();
            Console.Log("Start", "port-message-waiting/{0}", this);
        }

        //OnMessage = null;
        Console.Log("Start", "port-message-end/{0}", this);
    }

    private readonly HashSet<Signal> CloseSignal = new();
    public event Signal OnClose
    {
        add { if (value != null) lock (CloseSignal) { CloseSignal.Add(value); } }
        remove { lock (CloseSignal) { CloseSignal.Remove(value); } }
    }

    public Task Close()
    {
        return _cs.EmitClose();
    }

    private async Task _close()
    {
        /// 关闭输出就行了
        _channel_out.Complete();
        await CloseSignal.EmitAndClear();
        Console.Log("Close", "port-closed/{0}", this);
    }

    private readonly HashSet<Signal<I>> MessageSignal = new();
    public event Signal<I> OnMessage
    {
        add { if (value != null) lock (MessageSignal) { MessageSignal.Add(value); } }
        remove { lock (MessageSignal) { MessageSignal.Remove(value); } }
    }

    /**
     * <summary>
     * 发送消息，这个默认会阻塞
     * </summary>
     */
    public async Task PostMessageAsync(O msg)
    {
        Console.Log("PostMessage", "message-out/{0} >> {1}", this, msg);
        var success = await _channel_out.SendAsync(msg);

        if (!success)
        {
            Console.Log("PostMessage", "handle the closed channel case!");
        }
    }
}

public class NativeMessageChannel<T1, T2>
{
    /**
     * 默认锁住，当它解锁的时候，意味着通道关闭
     */
    private readonly SharedCloseSignal _cs = new();
    private BufferBlock<T1> _channel1 = new(new DataflowBlockOptions { BoundedCapacity = DataflowBlockOptions.Unbounded });
    private BufferBlock<T2> _channel2 = new(new DataflowBlockOptions { BoundedCapacity = DataflowBlockOptions.Unbounded });
    public NativePort<T1, T2> Port1;
    public NativePort<T2, T1> Port2;

    public NativeMessageChannel()
    {
        Port1 = new NativePort<T1, T2>(_channel1, _channel2, _cs);
        Port2 = new NativePort<T2, T1>(_channel2, _channel1, _cs);
    }
}